File: prog/level_1_2.py
# -------------------------------------------------------------------------------- 1. 체육복 시간초과 - 복습
# ------------------------------------------- 내꺼 25분 컷! - 시간초과 
# def solution(n, lost, reserve):
#   student = [0 for _ in range(n)]
#   for i in lost:
#     student[i-1] -= 1
#   for i in reserve:
#     student[i-1] += 1

#   print(student)
#   for i in range(n):
#     if i > 0 and student[i] == 1 and student[i-1] == -1:
#       student[i-1] = 0
#       student[i] = 0
#     elif i < n-1 and student[i] == 1 and student[i+1] == -1: 
#       student[i+1] = 0
#       student[i] = 0
#   print(student)

#   answer = 0
#   for i in student:
#     if i >= 0:
#       answer += 1
#   return answer

# print(solution(5, [2,4], [1,3,5]))
# print(solution(5, [2,4], [3]))
# print(solution(5, [2], [3]))
# print(solution(3, [3], [1]))
# print(solution(2, [0], [1]))
# print(solution(2, [1], [1]))

# -------------------------------------------------------------------------------- 2. 없는 숫자 더하기
# ------------------------------------------- 내꺼 3분 컷!
# def solution(numbers):
#   answer = []
#   for i in range(10):
#     if i not in numbers:
#       answer.append(i)
#   return sum(answer)

# print(solution([1,2,3,4,6,7,8,0]))
# print(solution([5,8,4,0,6,7,9]))

# -------------------------------------------------------------------------------- 3. 같은 숫자는 싫어 - 복습!
# ------------------------------------------- 내꺼 2번째 시도에서 5분 컷! 근데 이상하게 헤맸었다!
def solution(a):
  n = len(a)
  answer = [a[0]]
  for i in range(1,n):
    if a[i] != answer[-1]:
      answer.append(a[i])

  return answer


print(solution([1,1,3,3,0,1,1]))
print(solution([4,4,4,3,3]))
print(solution([4,4,4,4,4]))
print(solution([1,2]))

# -------------------------------------------------------------------------------- 4. 평균 구하기 
# ------------------------------------------- 내꺼 1분 컷!
# def solution(arr):
#   a = sum(arr)/len(arr)
#   return round(a, 2)

# print(solution([1, 4, 5]))
# print(solution([1,2,3,4]))
# print(solution([5,5]))

# -------------------------------------------------------------------------------- 5. 자릿수 더하기
# ------------------------------------------- 내꺼 3분 컷!
# def solution(n):
#   answer = sum(list(map(int,str(n))))
#   return answer

# print(solution(123))
# print(solution(987))

# -------------------------------------------------------------------------------- 6. 문자열 다루기 기본 - 시간초과!! 복습!
# ------------------------------------------- 내꺼 아 어이없게...ㅋㅋㅋㅋㅅㅂ 16분 컷
# def solution(s):
#   if len(s) == 4 or len(s) == 6:
#     for i in s:
#       if i.isdigit() == False:
#         return False
#     return True
#   return False

# ------------------------------------------- 커뮤니티ㅋㅋㅋㅋ
# def solution(s):
#   return len(s) in [4,6] and s.isdigit()

# -------------------------------------------------------------------------------- 7. 문자열을 정수로 바꾸기
# ------------------------------------------- 내꺼 7분 컷!
# def solution(s):
#   answer = 0
#   if s[0] == '+':
#     answer = int(s[1:])
#   elif s[0] == '-' :
#     answer = -int(s[1:])
#   else:
#     answer = int(s)
#   return answer

# ------------------------------------------- 커뮤니티 ㅋㅋㅋㅋㅋㅋㅋㅅㅂ 개쩌네
# def solution(s):
#   return int(s)

# print(solution('+1234'))
# print(solution('1234'))
# print(solution('-1234'))

# -------------------------------------------------------------------------------- 8. 소수 만들기
# ------------------------------------------- 내꺼 11분 컷!
# from math import sqrt
# from itertools import combinations


# def prime_num(n):
#   for i in range(2,int(sqrt(n)) + 1):
#     if n % i == 0:
#       return False
#   return True

# def solution(nums):
#   count = 0
#   num_lists = list(combinations(nums,3))
#   for num in num_lists:
#     if prime_num(sum(num)):
#       count += 1
#   return count


# print(solution([1,2,3,4]))
# print(solution([1,2,7,6,4]))

# -------------------------------------------------------------------------------- 9. 약수의 개수와 덧셈
# ------------------------------------------- 내꺼 11분 컷!
# from math import sqrt

# def solution(left, right):
#   answer = 0
#   for i in range(left, right + 1):
#     if int(sqrt(i)) != sqrt(i):
#       answer += i
#     else:
#       answer -= i
#   return answer

# print(solution(13,17))
# print(solution(24,27))

# -------------------------------------------------------------------------------- 10. 부족한 금액 계산하기
# ------------------------------------------- 내꺼 12분 컷!
# def solution(price, money, count):
#   pay = 0
#   for i in range(1, count+1):
#     pay += (price * i)
#   if money >= pay:
#     return 0
#   else:
#     return pay - money

# print(solution(3,20,4))

File: prog/javascript/level_1_2.js
'use strict';
// -------------------------------------------------------------------------------- 푼 문제
// 문자열 내 p와 y의 개수
// x만큼 간격이 있는 n개의 숫자
// 핸드폰 번호 가리기
// 콜라츠 추측
// 짝수와 홀수
// -----
// 제일 작은 수 제거하기
// 자연수 뒤집어 배열로 만들기
// 정수 내림차순으로 배치하기
// 하샤드 수
// 이상한 문자 만들기
// -----
// 평균 구하기
// 자릿수 더하기
// 최대공약수와 최소공배수
// 정수 제곱근 판별
// 수박수박수박수박수박수?
// -----
// 문자열 내림차순으로 배치하기
// 문자열 내 마음대로 정렬하기
// 서울에서 김서방 찾기
// 없는 숫자 더하기
// 부족한 금액 계산하기

// -------------------------------------------------------------------------------- API
// --------------------------------------------- unshift()
// --------------------------------------------- shift()
// --------------------------------------------- reduce()
// Array.reduce(함수)
// -------------------------------------------------------------------------------- 1. 평균 구하기
// ------------------------------------------- 내꺼
// function solution(arr) {
//   let sum = 0
//   for(let i in arr){
//     sum += arr[i]
//   }
//   return sum/arr.length
// }

// console.log(solution([1,2,3,4]))
// console.log(solution([5,5]))
// -------------------------------------------------------------------------------- 2. 자릿수 더하기
// ------------------------------------------- 내꺼
// function solution(n){
//   const stringN = n.toString();
//   let sum = 0;
//   for(let i in stringN){
//     sum += parseInt(stringN[i]);
//   }
//   return sum
// }

// console.log(solution(123));
// console.log(solution(987));

// -------------------------------------------------------------------------------- 3. 최대공약수와 최소공배수
// ------------------------------------------- 내꺼
// function solution(n, m) {
//   let gcd = 1;
//   for(let i = 2; i <= Math.min(n,m); i++){
//     if (n%i === 0 && m % i === 0){
//       gcd = i;
//     }
//   }
//   const lcm = (n*m)/gcd;

//   return [gcd,lcm];
// }

// console.log(solution(3,12));
// console.log(solution(2,5));
// console.log(solution(12,24));

// -------------------------------------------------------------------------------- 4. 정수 제곱근 판별
// ------------------------------------------- 내꺼
// function solution(n){
//   var root = Math.sqrt(n);
//   if((parseInt(root) - root) === 0){
//     return (root + 1 ) **2;
//   } else{
//     return -1;
//   }
// }

// console.log(solution(1))
// console.log(solution(121))
// console.log(solution(3))
// console.log(solution(16))

// -------------------------------------------------------------------------------- 5. 수박수박수박수박수박수?
// ------------------------------------------- 내꺼
// function solution(n) {
//   let answer = '';
//   for(let i = 1; i <= n; i++){
//     if (i%2 == 1){
//       answer += '수';
//     } else{
//       answer += '박';
//     }
//   }
//   return answer
// };

// console.log(solution(3));
// console.log(solution(4));

// -------------------------------------------------------------------------------- 6. 문자열 내림차순으로 배치하기
// ------------------------------------------- 내꺼
// function solution(s) {
//   const splitedS = s.split('');
//   const sortedS = splitedS.sort().reverse();
//   return sortedS.join('');
// };

// console.log(solution("Zbcdefg"));

// -------------------------------------------------------------------------------- 7. 문자열 내 마음대로 정렬하기
// ------------------------------------------- 내꺼
// function solution(strings, n) {
//   for(let i=0; i<strings.length; i++){
//     strings[i] = strings[i].split("");
//     strings[i].unshift(strings[i].splice(n,1).join(""));
//     strings[i] = strings[i].join('');
//   }
//   strings.sort();
//   for(let j=0; j<strings.length; j++){
//     strings[j] = strings[j].split("");
//     strings[j].splice(n,0,strings[j].shift());
//     strings[j] = strings[j].join('');
//   }

//   return strings;
// };

// console.log(solution(["sun", "bed", "car"],1))
// console.log(solution(["abce", "abcd", "cdx"],2))

// -------------------------------------------------------------------------------- 8. 서울에서 김서방 찾기
// ------------------------------------------- 내꺼
// function solution(seoul) {
//   const i = seoul.indexOf('Kim');
//   return `김서방은 ${i}에 있다.`
// };

// console.log(solution(["Jane", "Kim"]))

// -------------------------------------------------------------------------------- 9. 없는 숫자 더하기
// ------------------------------------------- 내꺼
// function solution(numbers) {
//   let answer = 0;
//   for(let i=0; i<10; i++){
//     if(numbers.indexOf(i) == -1){
//       answer += i;
//     };
//   };
//   return answer;
// }

// console.log(solution([1,2,3,4,6,7,8,0]))
// console.log(solution([5,8,4,0,6,7,9]))

// -------------------------------------------------------------------------------- 10. 부족한 금액 계산하기
// ------------------------------------------- 내꺼
// function solution(price, money, count) {
//   let sum = 0;
//   for(let i = 1; i <=count; i++){
//     sum += (price*i)
//   }
//   return money > sum? 0 : sum-money;
// };

// console.log(solution(3,20,4))